controllers/remote/cluster_cache.go

```diff
@@ -0,0 +1,401 @@
+// kind, and eventHandler, then this is a no-op.
+func (m *ClusterCacheTracker) Watch(ctx context.Context, input WatchInput) error {
+	wi := watchInfo{
+		watcher:      input.Watcher,
+		kind:         input.Kind,
+		eventHandler: input.EventHandler,
+	}
+
+	// First, check if the watch already exists
+	var exists bool
+	m.watchesLock.RLock()
+	exists = m.watchExists(input.Cluster, wi)
+	m.watchesLock.RUnlock()
+
+	if exists {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
+
+	// Doesn't exist - grab the write lock
+	m.watchesLock.Lock()
+	defer m.watchesLock.Unlock()
+
+	// Need to check if another goroutine created the watch while this one was waiting for the lock
+	if m.watchExists(input.Cluster, wi) {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
+
+	// Need to create the watch
+	watchesForCluster, found := m.watches[input.Cluster]
+	if !found {
+		watchesForCluster = make(map[watchInfo]struct{})
+		m.watches[input.Cluster] = watchesForCluster
+	}
+
+	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
+	if err != nil {
+		return err
+	}
+
+	if err := input.Watcher.Watch(source.NewKindWithCache(input.Kind, cache), input.EventHandler, input.Predicates...); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("error creating watch"))
+	}
+
+	watchesForCluster[wi] = struct{}{}
+
+	return nil
+}
+
+// getOrCreateClusterCache returns the clusterCache for cluster, creating a new ClusterCache if needed.
+func (m *ClusterCacheTracker) getOrCreateClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	cache := m.getClusterCache(cluster)
+	if cache != nil {
+		return cache, nil
+	}
+
+	return m.newClusterCache(ctx, cluster)
+}
+
+// getClusterCache returns the clusterCache for cluster, or nil if it does not exist.
+func (m *ClusterCacheTracker) getClusterCache(cluster client.ObjectKey) *clusterCache {
+	m.clusterCachesLock.RLock()
+	defer m.clusterCachesLock.RUnlock()
+
+	return m.clusterCaches[cluster]
+}
+
+// newClusterCache creates and starts a new clusterCache for cluster.
+func (m *ClusterCacheTracker) newClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
+	m.clusterCachesLock.Lock()
+	defer m.clusterCachesLock.Unlock()
+
+	// If another goroutine created the cache while this one was waiting to acquire the write lock, return that
+	// instead of overwriting it.
+	if c, exists := m.clusterCaches[cluster]; exists {
+		return c, nil
+	}
+
+	config, err := RESTConfig(ctx, m.client, cluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{Scheme: m.scheme})
```

Comment:
Plumb through CacheOptions. Need to decide on the scheme. We currently copy the Manager's scheme into ClusterCacheTracker.scheme when creating the tracker. We can either

1. Use CacheOptions.Scheme if present; fall back to m.scheme
1. Remove m.scheme entirely and always/only use CacheOptions.Scheme

I think 1 gives more flexibility but could be confusing (maybe?). If we choose this option, let's document it well.

Reply:
I'm +1 for option 2, it is more explicit